ethcontract/src/batch.rs

```diff
@@ -0,0 +1,89 @@
+//! Module containing components to batch multiple contract calls
+//! into a single request to the Node.
+
+use futures::channel::oneshot::{channel, Sender};
+use web3::{
+    error::Error as Web3Error,
+    helpers::{self},
+    types::{BlockId, BlockNumber, Bytes, CallRequest},
+    BatchTransport as Web3BatchTransport,
+};
+
+/// Struct allowing to batch multiple calls into a single Node request
+pub struct CallBatch<T: Web3BatchTransport> {
+    inner: T,
+    requests: Vec<(Request, CompletionHandler)>,
+}
+
+type Request = (CallRequest, Option<BlockId>);
+type CompletionHandler = Sender<Result<Bytes, Web3Error>>;
+
+impl<T: Web3BatchTransport> CallBatch<T> {
+    /// Create a new instance from a BatchTransport
+    pub fn new(inner: T) -> Self {
+        Self {
+            inner,
+            requests: Default::default(),
+        }
+    }
+
+    /// Adds a call request to the current batch. The resulting future can only resolve after
+    /// the batch has been resolved via `execute_all`.
+    /// Panics, if the batch is dropped before executing.
+    pub fn push(
+        &mut self,
+        call: CallRequest,
+        block: Option<BlockId>,
+    ) -> impl std::future::Future<Output = Result<Bytes, Web3Error>> {
+        let (tx, rx) = channel();
+        self.requests.push(((call, block), tx));
+        async move { rx.await.expect("Batch has been dropped without executing") }
```

Comment:
Instead of panicking it could block forever which is another reasonable result of the transport never having been executed. Or we could return an error. This is a little restrictive because we are using the web3 Error here but I see it does have a `Transport(String)` variant which we could repurpose.
It comes down to whether there can be a good reason to await a batch while not running the transport. Not sure.

Reply:
I was thinking about adding a special BatchError return type here which could also be returned in case we await on a single call before we await the batch, but decided against it as it's likely a programming error and requires less convenient handling in the calling code.

I'm not a huge fan of starving the future as this could lead to deadlocks and seems harder to debug. In this case I'd rather panic.

Reply:
> return type here which could also be returned in case we await on a single call before we await the batch

I think this is a legitimate use case: you batch a bunch of calls and create fibers to handle each call, the task scheduler decides to schedule one of these fibers before the one that calls `execute_all`.

Reply:
Yes, this was also the reason why I'm not panicking in this case (cf. description). 

Would you prefer an Error in the case the batch is dropped before the future is polled (I really don't like not resolving the future if we know that it will never be able to afterwards)?

Reply:
> (I really don't like not resolving the future if we know that it will never be able to afterwards

Shouldn't returning an `Err` here just resolve the `Future` to an error (as the `async { }` block is returning the error)?.

Reply:
My dislike for starvation is referring to the proposal of "it could block forever".

> Shouldn't returning an Err here just resolve the Future to an error

Yes, and it would change the function return type to `std::result::Result<std::result::Result<web3::types::Bytes, web3::Error>, Canceled>`. We can certainly do that, my concern was that it adds more inconvenient handling in the calling code (likely we will just unwrap there) as I don't really see why you would want to allow dropping the Batch before awaiting it. As such I felt it was ok to panic here.

Reply:
> Yes, and it would change the function return type to std::result::Result<std::result::Result<web3::types::Bytes, web3::Error>, Canceled>.

I like @vkgnosis solution of using `Transport(String)` as it _kind of_ fits:
```
rx.await.unwrap_or_else(|| Web3Error::Transport("Batch has been dropped without executing".to_owned())
```